planet/http.py

```diff
@@ -151,35 +150,50 @@ async def aclose(self):
     async def _retry(self, func, *a, **kw):
         """Run an asynchronous request function with retry.
 
+        Retry uses exponential backoff
+
         Raises:
             planet.exceptions.TooManyRequests: When retry limit is exceeded.
         """
-        # TODO: retry will be provided in httpx v1 [1] with usage [2]
-        # 1. https://github.com/encode/httpcore/pull/221
-        # 2. https://github.com/encode/httpx/blob/
-        # 89fb0cbc69ea07b123dd7b36dc1ed9151c5d398f/docs/async.md#explicit-transport-instances # noqa
+        # NOTE: if we need something more fancy, consider the following libs:
+        # * https://pypi.org/project/retry/
+        # * https://pypi.org/project/retrying/
+        # TODO: consider increasing the scope of retryable exceptions, ex:
+        # https://github.com/googleapis/python-storage/blob/1dc6d647b86466bf133
+        # fe03ec8d76c541e19c632/google/cloud/storage/retry.py#L23-L30
         # TODO: if throttling is necessary, check out [1] once v1
         # 1. https://github.com/encode/httpx/issues/984
-        retry_count = self.retry_count
-        wait_time = self.retry_wait_time
-
         num_tries = 0
         while True:
             num_tries += 1
             try:
                 resp = await func(*a, **kw)
                 break
             except exceptions.TooManyRequests as e:
-                if num_tries > retry_count:
+                if num_tries > MAX_RETRIES:
                     raise e
                 else:
                     LOGGER.debug(f'Try {num_tries}')
+                    wait_time = self._calc_exponential_wait(num_tries)
                     LOGGER.info(f'Too Many Requests: sleeping {wait_time}s')
-                    # TODO: consider exponential backoff
-                    # https://developers.planet.com/docs/data/api-mechanics/
                     await asyncio.sleep(wait_time)
         return resp
 
+    @staticmethod
+    def _calc_exponential_wait(num_tries):
+        """Calculates exponential wait
+
+        Introduces some randomness to stagger multiple requests and avoid
+        waves. Also thresholds maximum wait value to MAX_BACKOFF.
+
+        Ref:
+        * https://developers.planet.com/docs/data/api-mechanics/
+        * https://cloud.google.com/iot/docs/how-tos/exponential-backoff
+        """
+        random_number_milliseconds = random.randint(0, 1000) / 1000.0
+        calc_wait = 2**num_tries + random_number_milliseconds
```

Comment:
This is so smart, I love this! You mentioned you found this idea in some resource, which one was that?

Reply:
Its the resource referenced on line 191 :)